src/poll.rs

```diff
@@ -393,115 +717,202 @@ impl fmt::Debug for Registration {
+            }
         }
-    }
 
-    fn node(&self) -> &ReadinessNode {
-        self.node.as_ref().unwrap()
-    }
+        // Now enter the compare-and-swap loop
+        loop {
+            next = state;
 
-    fn registration_data_mut(&self, readiness_queue: &ReadinessQueue) -> io::Result<&mut RegistrationData> {
-        // `&Poll` is passed in here in order to ensure that this function is
-        // only called from the thread that owns the `Poll` value. This is
-        // required because the function will mutate variables that are read
-        // from a call to `Poll::poll`.
+            // The node is only dropped once all `Registration` handles are
+            // dropped. Only `Registration` can call `update`.
+            debug_assert!(!state.is_dropped());
 
-        if !self.queue.identical(readiness_queue) {
-            return Err(io::Error::new(io::ErrorKind::Other, "registration registered with another instance of Poll"));
+            // Update the write token position, this will also release the token
+            // to Poll::poll.
+            if curr_token != token {
+                next.set_token_write_pos(next_token_pos);
+            }
+
+            // Update readiness and poll opts
+            next.set_interest(interest);
+            next.set_poll_opt(opt);
+
+            // If there is effective readiness, the node will need to be queued
+            // for processing. This exact behavior is still TBD, so we are
+            // conservative for now and always fire.
+            //
+            // See https://github.com/carllerche/mio/issues/535.
+            if next.effective_readiness().is_some() {
+                next.set_queued();
+            }
+
+            // compare-and-swap the state values. Only `Release` is needed here.
+            // The `Release` ensures that `Poll::poll` will see the token
+            // update and the update function doesn't care about any other
+            // memory visibility.
+            let actual = self.state.compare_and_swap(state, next, Release);
+
+            if actual == state {
+                break;
+            }
+
+            // CAS failed, but `curr_token_pos` should not have changed given
+            // that we still hold the update lock.
+            debug_assert_eq!(curr_token_pos, actual.token_write_pos());
+
+            state = actual;
         }
 
-        Ok(self.node().registration_data_mut())
+        // Release the lock
+        self.update_lock.store(false, Release);
+
+        if !state.is_queued() && next.is_queued() {
+            // We are responsible for enqueing the node.
+            try!(self.queue.enqueue_node_with_wakeup(self));
+        }
+
+        Ok(())
+    }
+
+    /// Set the node's dropped flag, informing `Poll::poll` that it should
+    /// decrement its ref count.
+    fn flag_as_dropped(&self) {
+        self.state.flag_as_dropped();
+    }
+}
+
+impl ops::Deref for RegistrationInner {
+    type Target = ReadinessNode;
```

Comment:
It's mostly to not have to always have an unsafe block to deref the node. Not sure what you mean by "doesn't have any more inherent methods".